doc/design/proposal-ambient-loaders.md

```diff
@@ -0,0 +1,47 @@
+# Ambient Loaders Design
+
+## Problem
+
+Loaders let side logic be applied when resolving import statements. However, this only applies to the main application: loaders themselves aren’t currently affected by any other loaders. As a result, 3rd-party loaders cannot be injected if accessing them requires going through another loader.
+
+For instance, imagining a loader that would allow modules to be loaded from a `node_modules.zip` file (instead of the typical folder), the following wouldn’t work:
+
+```
+node --loader zip --loader ts-node ./my-tool.mts
+```
+
+Indeed, importing `ts-node` would require the `zip` loader to contribute to the resolution (so that the final path loaded is `$PROJECT/node_modules.zip/ts-node` instead of `$PROJECT/node_modules/ts-node`, which doesn't exist), but it’s not what happens today: both `zip` and `ts-node` are resolved by the default Node resolver, preventing `ts-node` from resolving.
+
+The reason for that is an attempt to keep loaders as isolated as possible, to allow followup improvements like reusing them across workers or scaling them up. If all loaders were influenced by prior loaders, they’d effectively coalesce into a single one, which would prevent such work. Still, the problem remains.
```

Comment:
Are we 100% certain that we cannot let regular loaders be influenced by prior loaders? I'm afraid that if ambient loaders seem more powerful than regular loaders, we will just end up with everyone always using `--ambient-loader` and forget about the other flag.

Reply:
I think we said no because it would require reentrancy, and our preliminary thoughts on this Ambient Loader concept was that it's the nuclear option so it would be chosen very infrequently (too easy to foot-gun, so only worth it when you really need it and really know what you're doing).

Reply:
I'd be curious to hear about reentrancy hazards that will crop up in the types of loaders that can be non-ambient.  E.g. a use-case where you don't need `--ambient-loader` and yet are doing something tricky enough that you're likely to footgun.  A simple source->source compiler such as a coffeescript loader wouldn't have any of these risks, I don't think.

Reply:
@arcanis If you feel comfortable explaining it, do you mind adding a section to the document about reentrancy and its hazards and what this means for loaders/ambient loaders? Like perhaps how those issues can be addressed (or not) via ambient loaders, etc. Or what if we only allowed just one ambient loader?

Reply:
As I mentioned in [this thread](https://github.com/nodejs/loaders/pull/63#discussion_r807132506), I'm not sure myself I see the problems with reentrancy (nor I see it avoidable for resolution in general, or particularly affected by ambient loaders), so I'd prefer if someone else could make a writeup.

Reply:
Who is concerned about these hypothetical reentrancy concerns?  From an outside perspective, it appears as if perhaps it was hypothesized but no one could think of any examples.

Reply:
> Who is concerned about these hypothetical reentrancy concerns? From an outside perspective, it appears as if perhaps it was hypothesized but no one could think of any examples.

@bmeck was the one who knew the most about them.

Reply:
I'm also concerned about reentrancy, but for me that doesn't block adding the proposal itself, just a concern about a later stage.